plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java

```diff
@@ -3950,6 +3970,25 @@ protected Answer execute(RebootCommand cmd) {
         }
     }
 
+    private void checkAndSetEnableSetupConfig(VirtualMachineMO vmMo, VirtualMachineTO virtualMachine) {
+        if (virtualMachine.isEnterHardwareSetup()) {
+            VirtualMachineBootOptions bootOptions = new VirtualMachineBootOptions();
+            VirtualMachineConfigSpec vmConfigSpec = new VirtualMachineConfigSpec();
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug(String.format("configuring VM '%s' to reboot into hardware setup menu.",virtualMachine.getName()));
+            }
+            bootOptions.setEnterBIOSSetup(virtualMachine.isEnterHardwareSetup());
+            vmConfigSpec.setBootOptions(bootOptions);
+            try {
+                if (!vmMo.configureVm(vmConfigSpec)) {
+                    throw new Exception("Failed to configure VM to boot into hardware setup menu: " + vmMo.getName());
+                }
+            } catch (Exception e) {
+                s_logger.error(String.format("failed to reconfigure VM '%s' to boot into hardware setup menu",virtualMachine.getName()),e);
```

Comment:
We are swallowing the exception here, should we try to stop the process instead?

Reply:
no, the rest of the reboot process can still proceed normally and it is an unlikely event that implies an error in Vsphere.

Reply:
ok, revisited this and yes it is a valid concern in view of the hard reset that follows in the calling method. I'll have to refactor this to handle a failure to set the parameter on the VM.

Reply:
Looks good. I can think of a way to test this:
What about adding a config key to control the delay as described here? (bootDelay param) https://vdc-download.vmware.com/vmwb-repository/dcr-public/b50dcbbf-051d-4204-a3e7-e1b618c1e384/538cf2ec-b34f-4bae-a332-3820ef9e7773/vim.vm.BootOptions.html
Will need testing but as it expects a long value, sending a negative delay may cause VM config to throw an exception

Reply:
:) next feature, but good to keep in mind!